### PR TITLE
Use the non-blocking IO enabled http.rb instead of net/http in watchers

### PR DESCRIPTION
**Pull request by msufa**
When using kubeclient in an actor based concurrent environment, such as Celuloid, we need a non blocking HTTP library to handle watching Kubernetes events. http.rb can be used with Celuloid::IO sockets, so that Celluloid actors remain responsive when awaiting cluster events.

I've tested with basic and ssl authentication options, both with and without the usage of Celluloid::IO. I've not tested thoroughly using bearer token, I only verified that the headers are set when making a request.

This resolves issue #140

### CODE REVIEW

**Comment by msufa** [comment]
Not sure how to handle the `Avoid parameter lists longer than 5 parameters.` rubocop restriction. Please let me know if you can live with it or how would you rather expose the socket options to the user.

**Comment by simon3z** [comment]
Thanks @msufa! Don't mind the travis error for now. I expect this to be a PR with a long life-cycle as I want to be sure that everything works as expected. I'll start to test it now but I want to run with it for a week or more to check it thoroughly.

**Comment by simon3z** [comment]
@msufa I see that you've done some generic refactoring (thanks!) to facilitate the introduction of non-blocking IO but that is good regardless, for example creating `read_stream`, etc.

If you split that out into its own patch (you can use this PR anyway) it would facilitate my work of reviewing (keeping refactoring and features separated).
Thanks!

**Comment by msufa** [comment]
@simon3z, I only extracted that part because I got a warning from Code Climate that the host method is too complex. I can stuff that code back in if it makes the PR clearer.

**Comment by simon3z** [comment]
> @simon3z, I only extracted that part because I got a warning from Code Climate that the host method is too complex. I can stuff that code back in if it makes the PR clearer.

@msufa no, on the contrary I am all in favor of that but in the scope of this PR it makes it harder to review for relevant changes. Just split it to another patch and we're good.

**Comment by simon3z** [comment]
This is working for my use-cases, @abonas can you review?

**Comment by abonas** [comment]
Didn't see that this comment was addressed: https://github.com/abonas/kubeclient/pull/142#issuecomment-161709995
Did I miss it?

File: lib/kubeclient/common.rb [hunk 1]
```diff
@@ -24,6 +24,10 @@ def initialize_client(
         password: nil,
         bearer_token: nil,
         bearer_token_file: nil
+      },
+      socket_options: {
```
Comment (abonas) [inline lib/kubeclient/common.rb]: I am probably missing something here, but why is there a need to do the socket_options hash initialization both here and in kubeclient.rb line 55? the code looks the same and kubeclient.rb calls common.rb anyway

Comment (msufa) [inline lib/kubeclient/common.rb]: I agree @abonas, I was just trying to be consistent with the way `ssl_options` & `auth_options` are handled. I can clean them all up either here or in a different PR, whichever you fancy.

Comment (abonas) [inline lib/kubeclient/common.rb]: > I agree @abonas, I was just trying to be consistent with the way ssl_options & auth_options are handled. I can clean them all up either here or in a different PR, whichever you fancy.

imo no need to handle it in this PR, a separate one can be totally fine. @simon3z ?
And @msufa , thanks for your work!

**Comment by msufa** [comment]
> Didn't see that this comment was addressed: #142 (comment)
Did I miss it?

@abonas, indeed this comment was addressed, please have a look here https://github.com/zendesk/kubeclient/commit/ac7ac64c53710e9367ee9e6f09305d5e62be2c5c
I'll wait until this PR is merged and revert that commit to bring back the refactoring if that's ok.

File: lib/kubeclient/common.rb (hunk 1 above)
Comment (abonas) [inline lib/kubeclient/common.rb]: I can't see the build details for some reason, but Travis does fail...

**Comment by abonas** [comment]
> @abonas, indeed this comment was addressed, please have a look here zendesk@ac7ac64
> I'll wait until this PR is merged and revert that commit to bring back the refactoring if that's ok.

@simon3z was the one doing the more thorough review of this PR given that he implemented the watches area, I'll let him decide on the approach here.

File: lib/kubeclient/common.rb (hunk 1 above)
Comment (msufa) [inline lib/kubeclient/common.rb]: Travis is failing on a Rubocop check, I mentioned it here https://github.com/abonas/kubeclient/pull/142#issuecomment-161314282

Comment (simon3z) [inline lib/kubeclient/common.rb]: > imo no need to handle it in this PR, a separate one can be totally fine. @simon3z ?

:+1: (separate PR is better)

**Comment by simon3z** [comment]
@msufa please rebase and squash `move parsing of the HTTP response back into the #each method` in `update readme to include info about non-blocking socket support`

File: lib/kubeclient/common.rb [hunk 2]
```diff
@@ -337,19 +342,18 @@ def net_http_options(uri)
       }
 
       if uri.scheme == 'https'
-        options.merge!(
-          use_ssl: true,
+        options[:ssl] = {
```
Comment (simon3z) [inline lib/kubeclient/common.rb]: @msufa why is this change required?

Comment (msufa) [inline lib/kubeclient/common.rb]: http.rb takes SSL options from an `ssl` sub hash under main options, see https://github.com/httprb/http/blob/1836ae8d7986d4ce39ab417bdfd20b9098800edc/lib/http/options.rb#L47

Furthermore, `use_ssl` is not required for http.rb, it's a net/http concept, see http://ruby-doc.org/stdlib-2.2.4/libdoc/net/http/rdoc/Net/HTTP.html#class-Net::HTTP-label-HTTPS

**Comment by msufa** [comment]
@simon3z PR rebased, commits squashed.

File: lib/kubeclient/watch_stream.rb [hunk 3]
```diff
@@ -1,54 +1,63 @@
...
 module Kubeclient
   module Common
     # HTTP Stream used to watch changes on entities
     class WatchStream
       def initialize(uri, http_options, format: :json)
         @uri = uri
-        @http = nil
-        @http_options = http_options.merge(read_timeout: nil)
+        @http_client = nil
+        @http_options = http_options
```
Comment (simon3z) [inline lib/kubeclient/watch_stream.rb]: @msufa you removed the `read_timeout` but apparently you haven't introduced anything similar for the new `http` gem, don't we need something like:

http://www.rubydoc.info/gems/http/0.9.8/HTTP/Timeout/Global

File: lib/kubeclient/watch_stream.rb [hunk 4]
```diff
@@ -1,54 +1,63 @@
...
 
       def each
         @finished = false
-        @http = Net::HTTP.start(@uri.host, @uri.port, @http_options)
 
-        buffer = ''
-        request = generate_request
+        @http_client = build_client
+        response = @http_client.request(:get, @uri, build_client_options)
+        unless response.code < 300
```
Comment (simon3z) [inline lib/kubeclient/watch_stream.rb]: @msufa how is this related to "Use non-blocking IO"? It's a good improvement but it should be evaluated separately, so if it's not part of this PR please remove it.

**Comment by simon3z** [comment]
@msufa I think that to pass rubocop you have to add in `.rubocop.yml`:

    Metrics/ParameterLists:
      Max: 8

`8` seems a good compromise to me. Feel free to suggest another value, cc @abonas 

Otherwise @msufa you'll have to comply to current metric which is `5` and refactor `initialize_client` (it would be too much confusing in this PR though).

**Comment by abonas** [comment]
> @msufa I think that to pass rubocop you have to add in .rubocop.yml:
> 
> Metrics/ParameterLists:
>   Max: 8
> 8 seems a good compromise to me. Feel free to suggest another value, cc @abonas
> 
> Otherwise @msufa you'll have to comply to current metric which is 5 and refactor initialize_client (it would be too much confusing in this PR though).

no objections to change rubocop.yml threshold

File: lib/kubeclient/watch_stream.rb (hunk 3 above)
Comment (msufa) [inline lib/kubeclient/watch_stream.rb]: @simon3z, `HTTP` does not time out by default so no explicit setup is needed:
 - https://github.com/httprb/http#timeouts
 - https://github.com/httprb/http/blob/v0.9.8/lib/http/options.rb#L10

File: lib/kubeclient/watch_stream.rb (hunk 4 above)
Comment (msufa) [inline lib/kubeclient/watch_stream.rb]: @simon3z, unfortunately `HTTP` does not have a fancy `Success` wrapper class and deals only with raw integer codes. In case of `net/http` the trick is to have all the 2xx codes inherit from `Net::HTTPSuccess`, see [here](https://github.com/ruby/ruby/blob/trunk/lib/net/http/responses.rb#L36). So in our case the only difference is that we treat 1xx responses as success as well. I can fix this if you think it makes a difference.

**Comment by msufa** [comment]
I've relaxed the rubocop parameter list restriction to 8.
I've also fixed the `http.rb` dependency to version 0.9.8 for now, as the Celluloid::IO support was temporarily removed from 1.0, see [here](https://github.com/httprb/http/wiki/Parallel-requests-with-Celluloid::IO) for details.
/cc @simon3z @abonas

**Comment by abonas** [comment]
thanks @msufa , @simon3z please review/merge

**Comment by msufa** [comment]
Bump...

**Comment by abonas** [comment]
@msufa I pinged @simon3z and asked him to confirm/merge the PR - given that he is more familiar with the modified area and led the review. sorry for the delays.

**Comment by abonas** [comment]
@msufa if you could clarify the following to me - as that wiki states, Celluloid is removed from 1.0 of http (it's not clear from the link why it was removed), and setting hardcoded http gem version dependency in kubeclient to have Celluloid support can create issues for some kubeclient users who require other versions of the http gem at the moment, and can prevent from upgrading the gem version to enjoy other bug fixes, etc. Your thoughts on the topic?

**Comment by msufa** [comment]
@abonas, please have a look here for more details on `Celluloid::IO` support in `http.rb`: https://github.com/httprb/http/issues/272
Although there are many other benefits from using `http.rb`, the most significant use case for us is to utilise the non-blocking IO in watchers, implemented as Celluloid Actors, which constantly observe the state of the cluster. We believe that this is something that other users of `kubeclient` could benefit from.
I don't see any crucial changes in `http.rb` when comparing v0.9.8 with v1.0.0. I've also asked for some more details about the future of Celluloid support. Ultimately, upgrading the dependency to a version other than v0.9.8 should not break any code, as the extra socket parameters should just be ignored by the http library. This would be totally transparent for `kubeclient` users which don't rely on Celluloid support.

**Comment by abonas** [comment]
> @abonas, please have a look here for more details on Celluloid::IO support in http.rb: httprb/http#272
> Although there are many other benefits from using http.rb, the most significant use case for us is to utilise the non-blocking IO in watchers, implemented as Celluloid Actors, which constantly observe the state of the cluster. We believe that this is something that other users of kubeclient could benefit from.
> I don't see any crucial changes in http.rb when comparing v0.9.8 with v1.0.0. I've also asked for some more details about the future of Celluloid support. Ultimately, upgrading the dependency to a version other than v0.9.8 should not break any code, as the extra socket parameters should just be ignored by the http library. This would be totally transparent for kubeclient users which don't rely on Celluloid support.

ack. thanks!
